cdk/visit/lambda_code/log_visit/log_visit.py

```diff
@@ -0,0 +1,196 @@
+            This should:
+            1. Check whether this user has visited before by looking for a
+            sentinel record in the table
+            2. Trigger a registration workflow if this is the first time for that user
+            3. Place a visit entry into the table
+            """
+        # return client error if no string params
+
+        HEADERS = {
+            'Content-Type': 'application/json',
+            'Access-Control-Allow-Headers': 'Content-Type',
+            'Access-Control-Allow-Origin': 'https://visit.cumaker.space',
+            'Access-Control-Allow-Methods': 'OPTIONS,POST,GET'
+        }
+
+        if (request is None):
+            return {
+                'headers': HEADERS,
+                'statusCode': 400,
+                'body': json.dumps({
+                    "Message": "Failed to provide parameters"
+                })
+            }
+
+        try:
+            # Get the username from the request body.
+            username = json.loads(request["body"])["username"]
+            location = ' '
+            try:
+                location = json.loads(request["body"])["location"]
+            except Exception as e:
+                exception_type, exception_value, exception_traceback = sys.exc_info()
+                traceback_string = traceback.format_exception(
+                    exception_type, exception_value, exception_traceback)
+                err_msg = json.dumps({
+                    "errorType": "MissingParameter",
+                    "errorMessage": "Missing parameter: location",
+                    "errorTrace": traceback_string
+                })
+                self.logger.warn(err_msg)
+
+            # Check if this user has registered before.
+            registration = self.checkRegistration(username)
+
+            # If the user is not in the system, send a registration link.
+            if registration == 0:
+                self.registrationWorkflow(username)
+                # One could consider setting res = some other number here in order to
+                # bring up a page That lets the user know in order to sign in they
+                # have to check their email and register with the Makerspace.
+
+                # Call Function
+            res = self.addVisitEntry(username, location)
+
+            # Send response
+            return {
+                'headers': HEADERS,
+                'statusCode': res
+            }
+
+        except Exception as e:
+            # Return exception with response
+            return {
+                'headers': HEADERS,
+                'statusCode': 500,
+                'body': json.dumps({
+                    'Message': str(e)
+                })
+            }
+
+
+def handler(request, context):
+    # This will be hit in prod, and will connect to the stood-up dynamodb
+    # and Simple Email Service clients.
+    return LogVisitFunction(None, None).handle_log_visit_request(request, context)
```

Comment:
Small performance tip:
We can create this function object during lambda's static initialization phase, thereby only instantiating our clients once when the function is initialized. We can then reuse the clients that we have created throughout the lifetime of our lambda's execution environment

See https://docs.aws.amazon.com/lambda/latest/operatorguide/static-initialization.html for more detailse.g.

Ex.
```
# Initialize once.
log_visit_function = LogVisitFunction(None, None)

# Lambda entry point which is executed on every lambda invocation
def handler(request, context):
    return log_visit_function.handle_log_visit_request(request, context)
```

For a little context, the way that lambda works when it is invoked:
1. Creates an "execution environment" which effectively creates an ephemeral container (lasting ~15 mins) capable of being invoked `n` times over those 15 minutes. 
2. Invokes the function by running your "handler" code
3. The next time the function is invoked it will attempt to reuse a "warm" execution environement

The initialization period of an "execution environment"  is an expensive process relative to invoking a function that already has a "warm" execution environment. You can look into this process by googling "lambda cold start."

Reply:
Sounds good! I went ahead and moved around the code to run during the initialization process.